### PR TITLE
introduce rspec and convert over draw_spec.rb

### PR DESCRIPTION
**Pull request by mockdeep**
This is a quick first pass. The `marshal` test wasn't running because it wasn't
properly named before. I tried for a bit to get it to pass, but that might be a
project in its own right.

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: d45922d5265d081a20bba58c5a454b19ac0c02ce
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at http://robots.thoughtbot.com/5-useful-tips-for-a-better-commit-message

---
This message was auto-generated by https://gitcop.com

**Comment by mockdeep** [comment]
One unpleasant side effect with this is that the `:test` task won't run if `:spec` fails, so that could be a little annoying until everything is converted over.

**Comment by ioquatix** [comment]
I think this is an awesome direction to move in. +1.

**Comment by mockdeep** [comment]
Hmm, anybody have any idea why this build is failing? It looks as if older versions of ImageMagick are no longer available...

**Comment by mockdeep** [comment]
Looks like they changed their patch level. This will depend on https://github.com/gemhome/rmagick/pull/177

**Comment by ioquatix** [comment]
I actually asked the ImageMagick team to NOT do this, i.e. remove existing binaries. If the file isn't in `http://www.imagemagick.org/download/releases` then they have messed up again.

**Comment by mockdeep** [comment]
Yeah it looks like they removed it. Makes for things to be a little inconsistent on our end... Would it make sense for us to mirror them somewhere?

**Comment by ioquatix** [comment]
I don't think so.. but they shouldn't be doing it !@#$!@#. I explicitly asked them that when they make an official release that that file sticks around permanently.

**Comment by mockdeep** [comment]
Are they using an automated build script that just deletes them? Maybe we can patch that, too...

**Comment by mockdeep** [comment]
This has been rebased, so hopefully it will pass...

**Comment by mockdeep** [comment]
Ugh, it looks like the marshal spec passes under some versions of ImageMagick, which fails the build because a `pending` spec is passing... I think I'm just going to put a hard `fail` in there.

**Comment by mockdeep** [comment]
Alright, this is passing. Alright if I merge?

File: spec/rmagick/draw_spec.rb [hunk 1]
```diff
@@ -0,0 +1,157 @@
+class Magick::Draw
+  def self._dummy_img_
+    @@_dummy_img_
+  end
+end
+
+describe Magick::Draw do
+  before(:each) { @draw = Magick::Draw.new }
```
Comment (linduxed) [inline spec/rmagick/draw_spec.rb]: You could use the `let!` RSpec helper method to achieve the same thing here.

Comment (mockdeep) [inline spec/rmagick/draw_spec.rb]: I try to avoid `let!`, though I guess this is effectively the same. Let me try `let`, if we prefer that over instance variables.

Comment (linduxed) [inline spec/rmagick/draw_spec.rb]: [I too prefer not to use `let` in general](http://linduxed.com/blog/2014/08/24/writing-more-readable-rspec-tests/), but in this case you've basically done the exact same thing as `let!`, in which case one could use that. Whether memoization is a necessary measure in this case is a different discussion.

Comment (mockdeep) [inline spec/rmagick/draw_spec.rb]: Turns out `let` worked without the `!`. I normally use `let`, but I can see the benefit of instance variables since they make the so-called "mystery guests" a little more apparent. We also use the [rom style](https://github.com/solnic/rom-mapper/tree/master/spec/unit/rom/mapper) spec structure, one spec file per method, which also reduces the overhead, since the `let`s only apply to the current method.

**Comment by mockdeep** [comment]
gah! ImageMagick switched up their versions again!

**Comment by mockdeep** [comment]
@vassilevsky @linduxed any idea why this is failing? It seems to be unrelated to my changes here.

https://travis-ci.org/gemhome/rmagick/jobs/57288500

**Comment by vassilevsky** [comment]
It looks like the test uncovered a real mismatch.

File: spec/rmagick/draw_spec.rb [hunk 2]
```diff
@@ -0,0 +1,157 @@
+class Magick::Draw
+  def self._dummy_img_
+    @@_dummy_img_
+  end
+end
+
+RSpec.describe Magick::Draw do
+  let(:draw) { Magick::Draw.new }
+
+  describe '#_test_dummy_img_' do
```
Comment (linduxed) [inline spec/rmagick/draw_spec.rb]: By using a `#` at the start of the method name, you're implying that it's an instance method, when it's a class method on `Magick::Draw`. Please, change the `#` to a `.`.

File: spec/rmagick/draw_spec.rb [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+      draw.align = Magick::CenterAlign
+      draw.undercolor = Magick::Pixel.from_color('green')
+      draw.kerning = 10.5
+      draw.interword_spacing = 3.75
+
+      draw.circle(20, 25, 20, 28)
+      dumped = nil
+      expect { dumped = Marshal.dump(draw) }.not_to raise_error
+      expect { Marshal.load(dumped) }.not_to raise_error
+      fail 'this spec fails on some versions of ImageMagick'
```
Comment (linduxed) [inline spec/rmagick/draw_spec.rb]: Why was this line added? This kind of logic was not present in the tests written with the old test framework.

**Comment by linduxed** [comment]
I posted the issue #186 to address the problem. What's confusing is that I got that to pop up during my work on #185.

What's weird about this is that I can't get that error to go red anymore. Apparently, I ran `bundle exec rake` enough times that it doesn't want to fail anymore. The Ruby versions on Travis that go red don't go red for me; I have no idea why this is the case.

I did get the error for a while though and I'll try to look a bit more on the code to figure out what could be causing this.

File: spec/rmagick/draw_spec.rb (hunk 2 above)
Comment (mockdeep) [inline spec/rmagick/draw_spec.rb]: Good catch, I'll get that fixed.

File: spec/rmagick/draw_spec.rb (hunk 3 above)
Comment (mockdeep) [inline spec/rmagick/draw_spec.rb]: The issue here is complicated. Basically, because the original test didn't start with `test_` it wasn't actually getting run. When I converted it to RSpec it didn't pass for me locally so I marked it as `pending` above. The problem with that is that RSpec still runs the test, and if it passes, it marks it as failed because it shouldn't be marked as `pending` if it passes. It does actually pass on some versions of `ImageMagick`, but not on others, so it fails the build. I added this `fail` as a temporary measure so it just fails on all of them, and the `pending` passes. I figured investigating the issues related to this were unrelated to the conversion to RSpec.

Comment (linduxed) [inline spec/rmagick/draw_spec.rb]: In that case, I propose that another issue is filed, referenced in this PR, that outlines this problem.

Comment (mockdeep) [inline spec/rmagick/draw_spec.rb]: Added a new issue.

**Comment by mockdeep** [comment]
Renamed spec group to `._dummy_img_`.

**Comment by mockdeep** [comment]
I updated the spec to use `skip` instead of `pending`. I think that just skips it entirely.

**Comment by mockdeep** [comment]
Any idea why we're getting "Stack consistency error"?

https://travis-ci.org/gemhome/rmagick/jobs/58983037

**Comment by vassilevsky** [comment]
Sometimes it's consistency error, sometimes it's something else, but this test fails very often. I think it has something to do with the way the trace proc is implemented in the C extension. Maybe it uses some old MRI C API and newer versions of MRI support it badly. Something along these lines.

**Comment by mockdeep** [comment]
@vassilevsky any idea what needs to happen in order to make it consistent? Also, can we just merge this then?

**Comment by u338steven** [comment]
FWIW, `trace_proc` has a `volatile` variable in the C extension. ([at rmimage.c:L15224](https://github.com/gemhome/rmagick/blob/a11f7026f895efb64d3c60b66950f764bd82a0bd/ext/RMagick/rmimage.c#L15224))
Possibly, it might cause that confusion.

> 3) “volatile” implementations may be buggy/inconsistent in some
compilers and architectures. RB_GC_GUARD is customizable for broken
systems/compilers without those without negatively affecting other
systems.

[Appendix E. RB_GC_GUARD to protect from premature GC](https://github.com/ruby/ruby/blob/trunk/doc/extension.rdoc#appendix-e-rb_gc_guard-to-protect-from-premature-gc)

**Comment by vassilevsky** [comment]
I rebased this onto the awesome fixes by @u338steven and now it's passing! 🎉
The failing tests are the usuals.

**Comment by vassilevsky** [comment]
Thanks @mockdeep! 👊You are the trailblazer of better testing! 🚢
